knimin/lib/data_access.py

```diff
@@ -827,6 +827,104 @@ def pulldown(self, barcodes, blanks):
 
         return metadata, failures
 
+    def explain_pulldown_failures(self, barcodes):
+        """Builds failure reason list for barcodes passed
+
+        Parameters
+        ----------
+        barcodes : list of str
+            Barcodes to explain failure for
+
+        Returns: list of tuples of str
+            list in the form [(barcode, reason), (barcode, reason), ...]
+        """
+        def update_failure(sql, all_bc, reason):
+            hold = [x[0] for x in
+                    self._con.execute_fetchall(sql, [tuple(all_bc)])]
+            remaining_bc = all_bc.difference(hold)
+            return [(bc, reason) for bc in hold], remaining_bc
+
+        failure_info = []
+        current_barcodes = set(barcodes)
+
+        # TEST ORDER HERE MATTERS!
+        # not an AG barcode
+        sql = """SELECT barcode
+                FROM ag.ag_kit_barcodes
+                WHERE barcode IN %s
+                UNION
+                SELECT barcode
+                FROM ag.ag_handout_barcodes
+                WHERE barcode IN %s"""
+        hold = {x[0] for x in
+                self._con.execute_fetchall(sql, [tuple(current_barcodes)] * 2)}
+        failure_info.extend((bc, 'Not an AG barcode') for bc in
+                            current_barcodes.difference(current_barcodes))
+        current_barcodes = hold
+        # No more unexplained, so done
+        if len(current_barcodes) == 0:
+            return failure_info
+
+        # handout barcode
+        sql = """SELECT barcode
+                 FROM ag.ag_handout_barcodes
+                 WHERE barcode IN %s"""
+        failures, current_barcodes = update_failure(
+            sql, current_barcodes, 'Withdrawn sample')
```

Comment:
how is this a withdrawn sample given that the query below is more explicit on being a withdrawn sample?